services/hanke-service/src/main/kotlin/fi/hel/haitaton/hanke/domain/Hanke.kt

```diff
@@ -56,5 +64,38 @@ data class Hanke(
     var polyHaitta: Haitta13? = null
     var tarinaHaitta: Haitta13? = null
 
+    /** Note: this can be null for two reasons; the field wasn't requested for, or there are no geometries for the hanke.
+     * See 'tilaOnGeometrioita' field.
+     */
     var geometriat: HankeGeometriat? = null
+
+    // --------------- State flags -------------------
+    var tilat: HankeTilat = HankeTilat()
+
+    fun updateStateFlags() {
+        updateStateFlagOnKaikkiPakollisetLuontiTiedot()
+        updateStateFlagTiedotLiikHaittaIndeksille()
+    }
+
+    fun updateStateFlagOnKaikkiPakollisetLuontiTiedot() {
+        // All mandatory fields have been given... (though their validity should be checked elsewhere)
+        //  and saveType is submit, not just draft?
+        tilat.onKaikkiPakollisetLuontiTiedot = !nimi.isNullOrBlank()
+                && !kuvaus.isNullOrBlank()
+                && (alkuPvm != null) && (loppuPvm != null)
+                && (vaihe != null) && (vaihe != Vaihe.SUUNNITTELU || suunnitteluVaihe != null)
```

Comment:
I would prefer a bit human readable expression here... maybe even separater method to answer if "vaihe" checks are OK? ...because this no needs paper and pen and some drawing to understand :)

Reply:
Will do something. I have to admit, it is a bit unclear even for me, takes a moment or two to convince myself the logic works.